### ISSUE TITLE
Socket Server Error When Running

### ISSUE THREAD

**Issue by AndreasDemenagas** [sufst/intermediate-server#34]
I was trying to research alternative methods to implement the GUI front-end (other than Qt), when I ran into an interesting issue. I have encountered an interesting error while running the socket server from `main.py`

#### Background

* I had `main.py` up and running from the command line. 
* On a new terminal window and while `main` was running I began `car_emulator.py`. 

This resulted in a bunch of frames being logged into the console. At the same time I opened a Browser window and headed to `localhost:19900` and although nothing showed up in the screen (since the sockets server doesn't render anything in the DOM), everything seemed to work just fine. 

#### Issue

I wanted to test the socket server on `localhost:19900` so I quickly made a JavaScript front-end with socket-io to test the server and make sure that all the data is coming through correctly. 

I connected to the same `localhost:19900` address from the WebSockets front-end however, I the WebSockets server never seemed to connect. Checking the status of the server connection it was always in the `Connecting` stage but the JS front-end never established a socket connection to sockets server. The endpoint I used for the connection was: `ws://localhost:19900`. 

At the same time, I looked at the terminal window running `main.py` and I saw the following errors: I have highlighted with red the status of the connections to the front-end, and with green the error the server is throwing. 

![Screenshot 2021-01-06 at 2 20 25 PM](https://user-images.githubusercontent.com/55033328/103770264-63eb1300-502e-11eb-8d55-a10ec8a9ce1d.png)

So, to simplify things, the question is that although the frames from the server are logged in the console when `car_emulator.py` and `main.py` are running, I can't seem to be able to "catch" these frames from the WebSockets connection in JS and instead the `Error decoding frames` error is thrown. Does anyone know why this is happening?
 
Just for reference here is the front-end code. The log should appear in the console when the server has established a connection but when running it nothing is shown. 

```javascript
const socket = io('ws://localhost:19900')

socket.on('connect', () => {
    console.log('Websocket connected');
});

socket.on('disconnect', () => {
    console.log('Websocket disconnected');
});

```

**Comment by Nathanrs97** [comment]
This behaviour is expected and I've written about it in the wiki. Heres the useful part in the [wiki](https://github.com/sufst/intermediate-server/wiki/The-Protocol-Handler-Class) that is related to this behaviour
> If you are not familiar with sockets (sockets are low-level network access APIs on an OS to the underlying OS network stack, not to be confused with WebSockets which is a high-level network communication API built on low-level sockets and HTML)

You are trying to connect to the server using a WebSocket (which uses HTML and is seen in the raw data in the red box) which is not supported. To connect to the server you need to use a raw TCP socket. 

The reason why WebSockets are not currently supported is that the overhead of using WebSockets does not make it suitable for a low-latency, high messages/second protocol. To implement a WebSocket data-link layer in the intermediate server, the link must have a lower-frequency exchange rate (in the low hundreds of ms). Moreover, WebSockets are most designed for HTML based content rather than raw messaging, so is not designed for our workload and should be avoided if a raw TCP socket is available. 

I hope this helps. If it comes to a WebSocket is the Only available socket type available then we can implement WebSockets, but I would rather avoid it if possible.

**Comment by AndreasDemenagas** [comment]
Oh okay okay I think I finally understand now 😂. Thanks for explaining. I managed to successfully connect to the sockets server from Node.js using the built-in `net` module used for raw TCP sockets. 

I still haven't figured how to listen to the `data_received` and `connection_made` callbacks from the client but I will look more into it.

**Comment by Nathanrs97** [comment]
Awesome. 

`data_received` is a `factory` callback, not a `Protocol` class call back. If you manually create a client or server `factory` (`create_client` or `create_server`)you can use the `data_received` callback. Otherwise, you use the `Protocol` class callbacks for each frame type.

**Comment by AndreasDemenagas** [comment]
First of all really thanks for copping with me and taking the time to help me out in all of this. Still ultra new to this web and networking stuff and I know I ask a ton of really stupid questions. 

I have tried the following by I still don't get the raw frames thrown in the console. How am I supposed to catch the raw PDUs from the TCP client? 

Following doesn't seem to run although the server spits the data.
```JavaScript
client.on('data', function (chunk) {
    // This console log is never called.
    console.log(`Data received from the server: ${chunk.toString()}.`);
});
```

**Comment by Nathanrs97** [comment]
I'm going to assume (if the assumption is wrong, do let me know) that you do not send an `AIPDU` to the server once you have made a connection to it? The server does not know the client that has connected is a Car or a GUI until the client sends an `AIPDU` stating it is a Car or GUI first. 

If this is the case, please build an `AIPDU` (the wiki link is [here](https://github.com/sufst/intermediate-server/wiki/Telemetry-Protocol#22-application-initial-pdu-aipdu) for details on how to make an `AIPDU`) and send it to the server (as a GUI client) and it should start sending data to you.

**Comment by AndreasDemenagas** [comment]
Just tried that with the following AIPDU: 

```javascript
const aipdu = {
    "header": {
        "frame_type": 1,
        "seq_num": 1,
        "epoch": Date.now
    },
    "client_type": 3,
    "sw_ver": 1,
    "client_name": "T"
};
```
Then I converted the JSON to a `Buffer` object (as this is what node expects to be written) and I seem to get the following error during the decoding process of on the server side. 

<img width="1079" alt="Screenshot 2021-01-07 at 12 48 05 AM" src="https://user-images.githubusercontent.com/55033328/103827831-80b13600-5082-11eb-967e-62916e6abb3a.png">

**Comment by Nathanrs97** [comment]
Could you please paste the log so I can see the whole buffer that server receives? (your picture is cutting it off, the lines with the "->" in them)

**Comment by Nathanrs97** [comment]
I think the problem may be in your header definition. Please could you try:
```javascript
const aipdu = {
  "frame_type": 1,
  "seq_num": 1,
  "epoch": Date.now,
  "client_type": 3,
  "sw_ver": 1,
  "client_name": "T"
};
```

**Comment by AndreasDemenagas** [comment]
Here you go. 

![Screenshot 2021-01-07 at 12 57 30 AM](https://user-images.githubusercontent.com/55033328/103830021-b99dda80-5083-11eb-8739-13d0de68122e.png)

**Comment by AndreasDemenagas** [comment]
> I think the problem may be in your header definition. Please could you try:

Tried that, still doesn't work but completely different error. 

![Screenshot 2021-01-07 at 1 04 31 AM](https://user-images.githubusercontent.com/55033328/103831272-68dab180-5084-11eb-942f-b73ecdc6ff82.png)

**Comment by Nathanrs97** [comment]
Hmm I think the keys are Enums and not strings. Could you please try:
```javascript
const aipdu = {
  "1": 1,
  "2": 1,
  "3": Date.now,
  "5": 3,
  "6": 10000,
  "7": "GUI"
};
```

**Comment by AndreasDemenagas** [comment]
Okay I tried this and it gave me an error on key 3,  the `epoch`, so I assumed it couldn't decode `Date.now` into the format that it wanted. So, I replaced key 3 with just a random number and it finally worked!!! 

After running `main.py` and `car_emulator.py` I was able to catch the frames from the front-end!!!

By the way, here is the AIPDU I used just for reference. 

```javascript
const aipdu = {
    "1": 1,
    "2": 1,
    "3": 23, // Random number to replace Date.now that seemed to not work 😂
    "5": 3,
    "6": 10000,
    "7": "GUI"
};
```

I think a good idea would be to compile a running guide for the rest of the GUI team to have as guidance when they want to test the server. I can probably do that during the weekend. 

Thanks again for taking the time and helping me out
